src/native/libs/System.Security.Cryptography.Native.Android/pal_rsa.c

```diff
@@ -93,21 +139,71 @@ PALEXPORT int32_t AndroidCryptoNative_RsaPrivateDecrypt(int32_t flen, uint8_t* f
     if (!rsa->privateKey)
         return RSA_FAIL;
 
+    if (padding < Pkcs1 || padding > OaepSHA512)
+        return RSA_FAIL;
+
     abort_if_invalid_pointer_argument (to);
     abort_if_invalid_pointer_argument (from);
 
     JNIEnv* env = GetJNIEnv();
-
     jobject algName;
+    INIT_LOCALS(oaepLocals, oaepParameterSpec, oaepDigest, mgf1, mgf1ParameterSpec, pSource);
+
     if (padding == Pkcs1)
-        algName = make_java_string(env, "RSA/ECB/PKCS1Padding"); // TODO: is ECB needed here?
-    else if (padding == OaepSHA1)
-        algName = make_java_string(env, "RSA/ECB/OAEPPadding");
+    {
+        algName = make_java_string(env, "RSA/ECB/PKCS1Padding");
+    }
     else
-        algName = make_java_string(env, "RSA/ECB/NoPadding");
+    {
+        algName = make_java_string(env, "RSA/ECB/OAEPPadding");
+        oaepLocals[mgf1] = make_java_string(env, "MGF1");
+        oaepLocals[pSource] = (*env)->GetStaticObjectField(env, g_PSourcePSpecifiedClass, g_PSourcePSpecified_DefaultField);
+
+        if (padding == OaepSHA1)
+        {
+            oaepLocals[mgf1ParameterSpec] = (*env)->GetStaticObjectField(env, g_MGF1ParameterSpecClass, g_MGF1ParameterSpec_SHA1Field);
+            oaepLocals[oaepDigest] = make_java_string(env, "SHA-1");
+        }
+        else if (padding == OaepSHA256)
+        {
+            oaepLocals[mgf1ParameterSpec] = (*env)->GetStaticObjectField(env, g_MGF1ParameterSpecClass, g_MGF1ParameterSpec_SHA256Field);
+            oaepLocals[oaepDigest] = make_java_string(env, "SHA-256");
+        }
+        else if (padding == OaepSHA384)
+        {
+            oaepLocals[mgf1ParameterSpec] = (*env)->GetStaticObjectField(env, g_MGF1ParameterSpecClass, g_MGF1ParameterSpec_SHA384Field);
+            oaepLocals[oaepDigest] = make_java_string(env, "SHA-384");
+        }
+        else if (padding == OaepSHA512)
+        {
+            oaepLocals[mgf1ParameterSpec] = (*env)->GetStaticObjectField(env, g_MGF1ParameterSpecClass, g_MGF1ParameterSpec_SHA512Field);
+            oaepLocals[oaepDigest] = make_java_string(env, "SHA-512");
+        }
+        else
+        {
+            (*env)->DeleteLocalRef(env, algName);
+            RELEASE_LOCALS(oaepLocals, env);
+            return RSA_FAIL;
```

Comment:
It would be nice if the two versions of this can be shared.  But if the JNI tracking makes it hard, no worries.

Reply:
I think that this should be sharable without much trouble (and since the locals for this section are already in a separate array, it should be pretty easy to refactor)

Reply:
Okay. I think that is "right" and modeled the helper function off of [`GetMessageDigestInstance`](https://github.com/dotnet/runtime/blob/7414af2a5f6d8d99efc27d3f5ef7a394e0b23c42/src/native/libs/System.Security.Cryptography.Native.Android/pal_evp.c#L29) with regard to JNI lifetime things.